services/MerchStoreService.ts

```diff
@@ -100,47 +100,70 @@ export default class MerchStoreService {
+  private verifyVariantAndOptionConsistency(item: MerchItem | MerchandiseItemModel) {
+    if (!item.hasVariantsEnabled && item.options.length > 1) {
+      throw new UserError('Items with variants disabled cannot have multiple options');
+    }
+    if (item.hasVariantsEnabled && this.hasMultipleOptionTypes(item.options)) {
+      throw new UserError('Item cannot have multiple option types');
+    }
+  }
+
+  private hasMultipleOptionTypes(options: MerchItemOption[]) {
+    const optionTypes = new Set(options.map((option) => option.metadata.type));
+    return optionTypes.size > 1;
+  }
+
   public async editItem(uuid: Uuid, itemEdit: MerchItemEdit): Promise<MerchandiseItemModel> {
     return this.transactions.readWrite(async (txn) => {
       const merchItemRepository = Repositories.merchStoreItem(txn);
       const item = await merchItemRepository.findByUuid(uuid);
       if (!item) throw new NotFoundError();
       const { options, collection: updatedCollection, ...changes } = itemEdit;
 
-      const merchItemOptionRepository = Repositories.merchStoreItemOption(txn);
-      const updatedOptions = await Promise.all(options.map(async (optionUpdate) => {
-        const option = await merchItemOptionRepository.findByUuid(optionUpdate.uuid);
-        if (!option) throw new NotFoundError('Item option not found');
-        // 'quantity' is incremented instead of directly set to avoid concurrency issues with orders
-        // e.g. there's 10 of an item and someone adds 5 to stock while someone else orders 1
-        // so the merch store admin sets quantity to 15 but the true quantity is 14
-        if (optionUpdate.quantity) optionUpdate.quantity += option.quantity;
-        return merchItemOptionRepository.upsertMerchItemOption(option, optionUpdate);
-      }));
+      if (options) {
+        const updatedOptionMap = new Map(options.map((option) => [option.uuid, option]));
+        item.options.map((option) => {
+          if (!updatedOptionMap.has(option.uuid)) return;
+          const updatedOption = updatedOptionMap.get(option.uuid);
+          // 'quantity' is incremented instead of directly set to avoid concurrency issues with orders
+          // e.g. there's 10 of an item and someone adds 5 to stock while someone else orders 1
+          // so the merch store admin sets quantity to 15 but the true quantity is 14
+          if (updatedOption.quantity) updatedOption.quantity += option.quantity;
+          return MerchandiseItemOptionModel.merge(option, updatedOption);
+        });
+      }
 
-      const updatedOptionsIds = new Set(updatedOptions.map((option) => option.uuid));
-      item.options = [...updatedOptions, ...item.options.filter((option) => !updatedOptionsIds.has(option.uuid))];
+      // ensure the hasVariantsEnabled <--> options.length invariant is still consistent
+      if (changes.hasVariantsEnabled !== undefined) item.hasVariantsEnabled = changes.hasVariantsEnabled;
+      this.verifyVariantAndOptionConsistency(item);
```

Comment:
Let's just merge the item here and validate the new version.

```suggestion
      this.verifyItemHasValidOptions(MerchandiseItemModel.merge(item, changes));
```